java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,403 @@
+     * this [ArcHostContext].
+     */
+    protected open suspend fun maybeCancelResurrection(context: ArcHostContext) {
+        // TODO: wire up resurrection cancellation
+    }
+
+    /**
+     * Given a handle name, a [HandleConnection], and a [HandleHolder] construct an Entity
+     * [Handle] of the right type.
+     */
+    private suspend fun createHandle(
+        handleName: String,
+        handleSpec: Plan.HandleConnection,
+        holder: HandleHolder
+    ) = when (handleSpec.type) {
+        is SingletonType<*> ->
+            entityHandleManager.createSingletonHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        is CollectionType<*> ->
+            entityHandleManager.createSetHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        else -> throw IllegalArgumentException("Unknown type ${handleSpec.type}")
     }
 
     override suspend fun stopArc(partition: Plan.Partition) {
-        // TODO: not implemented yet
+        val arcId = partition.arcId
+        val context = lookupOrCreateArcHostContext(partition)
+        runningArcs[partition.arcId]?.let { context ->
+            when (context.arcState) {
+                ArcState.Running -> stopArcInternal(arcId, context)
+                ArcState.NeverStarted -> stopArcError(context, "Arc $arcId was never started")
+                ArcState.Stopped -> stopArcError(context, "Arc $arcId already stopped")
+                ArcState.Deleted -> stopArcError(context, "Arc $arcId is deleted.")
+            }
+        }
+    }
+
+    /**
+     * If an attempt to [ArcHost.stopArc] fails, this method should report the error message.
+     * For example, throw an exception or log.
+     */
+    private suspend fun stopArcError(context: ArcHostContext, message: String) {
+        // TODO: decide how to propagate this
+    }
+
+    /**
+     * Stops an [Arc], stopping all running [Particle]s, cancelling pending resurrection requests,
+     * releasing [Handle]s, and modifying [ArcState] and [ParticleState] to stopped states.
+     */
+    private suspend fun stopArcInternal(arcId: String, context: ArcHostContext) {
+        context.particles.values.forEach { particleContext ->
+            stopParticle(particleContext)
+        }
+        maybeCancelResurrection(context)
+        context.arcState = ArcState.Stopped
+        updateArcHostContext(arcId, context)
+    }
+
+    /**
+     * Shuts down a [Particle] by invoking its shutdown lifecycle methods, moving it to a
+     * [ParticleState.Stopped], and releasing any used [Handle]s.
+     */
+    private suspend fun stopParticle(context: ParticleContext) {
+        try {
+            context.particle.onShutdown()
+        } catch (e: Exception) {
+            // TODO: Shutdown failed, how to handle?
```

Comment:
Log, at the very least. To be even safer, rethrow, until you have a better idea what to do.